File: days/src/main/java/aoc2024j/day3/Day3.java

package aoc2024j.day3;

import utils.IO;

import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Gatherers;

public class Day3 {

    static final Pattern part1Pattern = Pattern.compile("mul\\((\\d{1,3}),(\\d{1,3})\\)");

    long sumLine(String line) {
        var matcher = part1Pattern.matcher(line).results();
        return matcher
                .mapToLong(m -> Long.parseLong(m.group(1)) * Long.parseLong(m.group(2)))
                .sum();
    }

    long part1(List<String> data) {
        return data.stream()
                .mapToLong(this::sumLine)
                .sum();
    }

    // make a regexp that matches mul(xxx,yyy) or do() or don't()

    static final Pattern part2Pattern = Pattern.compile(
            "mul\\((\\d{1,3}),(\\d{1,3})\\)|do\\(\\)|don't\\(\\)");

    record State(long result, boolean doMul) {
    }

    State sumLine2(State state, String line) {
        var matcher = part2Pattern.matcher(line).results();
        return matcher.gather(
                Gatherers.fold(
                        () -> state,
                        (s, m) -> {
                            if (m.group().startsWith("mul")) {
                                if (s.doMul) {
                                    return new State(
                                            s.result + Long.parseLong(m.group(1)) * Long.parseLong(m.group(2)),
                                            s.doMul);
                                } else {
                                    return s;
                                }
                            } else if (m.group().startsWith("don't")) {
                                return new State(s.result, false);
                            } else {
                                return new State(s.result, true);
                            }
                        }
                )).findFirst().orElseThrow();
    }

    long part2(List<String> data) {
        return data.stream()
                .gather(
                        Gatherers.fold(
                                () -> new State(0, true),
                                this::sumLine2
                        ))
                .findFirst()
                .orElseThrow()
                .result;
    }

    public static void main(String[] args) {
        var day3 = new Day3();
        var data = IO.getResourceAsList("aoc2024/day3.txt");
        var part1 = day3.part1(data);
        System.out.println("part1 = " + part1);
        var part2 = day3.part2(data);
        System.out.println("part2 = " + part2);
    }
}

File: days/src/main/scala/aoc2024s/day3/Day3.scala

package aoc2024s.day3;

import utils.IO;
import scala.jdk.CollectionConverters.*

object Day3 {

  val pattern1 = "mul\\((\\d{1,3}),(\\d{1,3})\\)".r

  def sumLine(line: String): Long = {
    pattern1.findAllMatchIn(line).map(m => m.group(1).toLong * m.group(2).toLong).sum
  }

  def part1(data: List[String]): Long = {
    data.map(sumLine).sum
  }

  val pattern2 = "mul\\((\\d{1,3}),(\\d{1,3})\\)|do\\(\\)|don't\\(\\)".r

  case class State(result: Long, doMul: Boolean) {
    def updated(arg1: Long, arg2: Long): State =
      if doMul then copy(result = result + arg1 * arg2) else this
  }

  def part2(data: List[String]): Long = {
    data.foldLeft(State(0, true)) { (state, line) =>
      pattern2.findAllMatchIn(line).foldLeft(state) { (state, m) =>
        m.group(0) match {
          case "do()" => state.copy(doMul = true)
          case "don't()" => state.copy(doMul = false)
          case _ => state.updated(m.group(1).toLong, m.group(2).toLong)
        }
      }
    }.result

  }

  @main def main3(): Unit = {
    val data = IO.getResourceAsList("aoc2024/day3.txt").asScala.toList;
    val part1 = Day3.part1(data);
    println(s"part1 = $part1");
    val part2 = Day3.part2(data);
    println(s"part2 = $part2");
  }
}